aten/src/ATen/native/mps/operations/LinearAlgebra.mm

```diff
@@ -91,6 +94,114 @@ void prepare_matrices_for_broadcasting(const Tensor* bias,
 
 enum LinearAlgebraOpType { ADDBMM_OP_TYPE, BADDBMM_OP_TYPE };
 
+void linalg_lu_factor_out_mps_impl(const Tensor& A, bool pivot, Tensor& LU, Tensor& pivots) {
+  using namespace mps;
+
+  Tensor A_t = A;
+  uint64_t aRows = A_t.size(-2);
+  uint64_t aCols = A_t.size(-1);
+  uint64_t aElemSize = A_t.element_size();
+  uint64_t numPivots = std::min(aRows, aCols);
+  resize_output(LU, A_t.sizes());
+  std::vector<int64_t> pivot_sizes(A_t.sizes().begin(), A_t.sizes().end() - 2);
+  pivot_sizes.push_back(numPivots);
+  resize_output(pivots, pivot_sizes);
+
+  if (A_t.numel() == 0 || LU.numel() == 0) {
+    LU.zero_();
+    return;
+  }
+
+  Tensor A_ = A_t.dim() > 3 ? A_t.flatten(0, -3) : A_t;
+  if (!A_.is_contiguous()) {
+    A_ = A_.clone(at::MemoryFormat::Contiguous);
+  }
+
+  uint64_t batchSize = A_.dim() > 2 ? A_.size(0) : 1;
+
+  std::vector<Tensor> status_tensors;
+  std::vector<Tensor> pivots_list;
+
+  status_tensors.reserve(batchSize);
+  pivots_list.reserve(batchSize);
+  for (C10_UNUSED const auto i : c10::irange(batchSize)) {
+    status_tensors.push_back(at::zeros(1, kInt, c10::nullopt, kMPS, c10::nullopt));
+    pivots_list.push_back(at::zeros(numPivots, kInt, c10::nullopt, kMPS, c10::nullopt));
+  }
+
+  // Since the LUDecomposition functions in-place if the result matrix completely aliases the source matrix,
+  // We copy LU from A as the new A if there is only one batch dim.
+  if (A_t.dim() <= 3) {
+    A_ = LU.copy_(A_);
+  }
```

Comment:
For `A_t.dim() > 3` cases we should clone `A_`; otherwise the input will be modified in-place. (Removing the above block:
```python
if (!A_.is_contiguous()) {
    A_ = A_.clone(at::MemoryFormat::Contiguous);
  }
```
)